src/distributions/adopt/installer.ts

```diff
@@ -0,0 +1,145 @@
+import * as core from '@actions/core';
+import * as tc from '@actions/tool-cache';
+
+import fs from 'fs';
+import path from 'path';
+import semver from 'semver';
+
+import { JavaBase } from '../base-installer';
+import { IAdoptAvailableVersions } from './models';
+import { JavaInstallerOptions, JavaDownloadRelease, JavaInstallerResults } from '../base-models';
+import { MACOS_JAVA_CONTENT_POSTFIX } from '../../constants';
+import { extractJdkFile, getDownloadArchiveExtension, isVersionSatisfies } from '../../util';
+
+export class AdoptDistribution extends JavaBase {
+  constructor(installerOptions: JavaInstallerOptions) {
+    super('Adopt', installerOptions);
+  }
+
+  protected async findPackageForDownload(version: string): Promise<JavaDownloadRelease> {
+    const availableVersionsRaw = await this.getAvailableVersions();
```

Comment:
Do understand this correctly? Are you downloading the _entire_ available list of AdoptOpenJDK binaries just to later filter if the version in the action matches one of the binaries?
If so, this is insane. The API provides functionality to e.g. get the latest 11-ga binary. Use this, or this action (or the entire GitHub IP range) will likely get banned for too many requests.

Reply:
Yes, it is necessary to download entire list to resolve customer's input correctly.
It is not enough to get `latest 11-ga binary` because user can specify version as `11` or `11.0` or `8.0.232` or even `8.0.282+8`. This task supports [SemVer](https://semver.org) notation for input so we need the full list of available versions to resolve input properly.

Actually, getting all versions is not a problem. Adopt just provides about ~50 versions that is pretty small list. It takes less than 0.1 seconds to retrieve it. It is not a problem even for Azul that provides wider list of available versions.

We don't see any problems with it

Reply:
> Yes, it is necessary to download entire list to resolve customer's input correctly.
> It is not enough to get `latest 11-ga binary` because user can specify version as `11` or `11.0` or `8.0.232` or even `8.0.282+8`. This task supports [SemVer](https://semver.org) notation for input so we need the full list of available versions to resolve input properly.

I would expect that the SemVer version of setup-java is parsed and converted into the version the respective distribution API expects. At least for the most common use-cases of `8` and `11` there should be no more than one request. You can always fallback to getting a wider range of versions if the action input is not sufficiently concise.

> Actually, getting all versions is not a problem. Adopt just provides about ~50 versions that is pretty small list. It takes less than 0.1 seconds to retrieve it. It is not a problem even for Azul that provides wider list of available versions.
> 
> We don't see any problems with it

Not yet.

And even if downloading the entire list is usually fast and works most of the time, an entire build would fail when just one of the paged (and unnecessery) requests fails - something that can be avoided for most use-cases.